server/src/main/java/org/opensearch/action/admin/indices/tiering/TieringRequestContext.java

```diff
@@ -0,0 +1,230 @@
+        for (Index index : acceptedIndices) {
+            indexTieringStatusMap.put(index, new IndexTieringInfo(index));
+        }
+        for (Map.Entry<Index, String> entry : failedIndices.entrySet()) {
+            indexTieringStatusMap.put(entry.getKey(), new IndexTieringInfo(entry.getKey(), IndexTieringState.FAILED, entry.getValue()));
+        }
+    }
+
+    public ActionListener<HotToWarmTieringResponse> getListener() {
+        return actionListener;
+    }
+
+    public IndexTieringInfo getIndexTieringInfo(Index index) {
+        return indexTieringStatusMap.get(index);
+    }
+
+    public boolean hasIndex(Index index) {
+        return indexTieringStatusMap.containsKey(index);
+    }
+
+    public Map<Index, String> getFailedIndices() {
+        Map<Index, String> failedIndicesMap = new HashMap<>();
+        for (IndexTieringInfo indexTieringInfo : filterIndicesByState(IndexTieringState.FAILED)) {
+            failedIndicesMap.put(indexTieringInfo.getIndex(), indexTieringInfo.getReason());
+        }
+        return failedIndicesMap;
+    }
+
+    public boolean isRequestProcessingComplete() {
+        return filterIndicesByState(IndexTieringState.SUCCESSFUL).size() + filterIndicesByState(IndexTieringState.FAILED)
+            .size() == indexTieringStatusMap.size();
+    }
+
+    public List<IndexTieringInfo> filterIndicesByState(IndexTieringState state) {
+        return indexTieringStatusMap.values()
+            .stream()
+            .filter(indexTieringInfo -> indexTieringInfo.getState() == state)
+            .collect(Collectors.toList());
+    }
+
+    public List<IndexTieringInfo> getIndicesPendingTiering() {
+        return indexTieringStatusMap.values()
+            .stream()
+            .filter(indexTieringInfo -> indexTieringInfo.getState() == IndexTieringState.IN_PROGRESS)
+            .collect(Collectors.toList());
+    }
+
+    @Override
+    public String toString() {
+        return "TieringRequestContext{" + "actionListener=" + actionListener + ", indexTieringStatusMap=" + indexTieringStatusMap + '}';
+    }
+
+    /**
+     * Represents info of a tiering index
+     *
+     * @opensearch.experimental
+     */
+    @ExperimentalApi
+    public static class IndexTieringInfo {
+        private final Index index;
+        private IndexTieringState state;
+        private String reason;
+
+        public IndexTieringInfo(Index index) {
+            this.index = index;
+            this.state = IndexTieringState.IN_PROGRESS;
+        }
+
+        public IndexTieringInfo(Index index, IndexTieringState state, String reason) {
+            this.index = index;
+            this.state = state;
+            this.reason = reason;
+        }
+
+        public Index getIndex() {
+            return index;
+        }
+
+        public IndexTieringState getState() {
+            return state;
+        }
+
+        public void markTiered() {
+            this.state = IndexTieringState.TIERED;
+        }
```

Comment:
looks like a state machine, do we need to validate the transitions are happening in the correct sequence?